internal/pkg/file/atomic.go

```diff
@@ -0,0 +1,341 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package file
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"io/fs"
+	"os"
+	"path/filepath"
+	"runtime"
+)
+
+// The internal options for atomic file writes.
+type atomicOpts struct {
+	target      string
+	permissions fs.FileMode
+	uid, gid    int
+}
+
+func (o *atomicOpts) wantsChmod() bool {
+	return o.permissions.IsRegular()
+}
+
+type AtomicOpener struct{ atomicOpts }
+
+// Prepares to open a new [Atomic] for the file at the given target path.
+func AtomicWithTarget(target string) *AtomicOpener {
+	return &AtomicOpener{atomicOpts{
+		target:      target,
+		permissions: fs.ModeIrregular, // use this as an "unset" marker, see wantsChmod()
+		uid:         -1,
+		gid:         -1,
+	}}
+}
+
+// The desired permissions for the target.
+// Will rely on the umask if not called.
+func (o *AtomicOpener) WithPermissions(perm os.FileMode) *AtomicOpener {
+	o.permissions = perm.Perm()
+	return o
+}
+
+// The desired owner UID for the target file.
+// Will be owned by the current user if not called.
+// Will have no effect on Windows.
+func (o *AtomicOpener) WithOwner(uid int) *AtomicOpener {
+	o.uid = max(-1, uid)
+	return o
+}
+
+// The desired group ID for the target file.
+// Will be owned by the current user's group if not called.
+// Will have no effect on Windows.
+func (o *AtomicOpener) WithGroup(gid int) *AtomicOpener {
+	o.gid = max(-1, gid)
+	return o
+}
+
+// Open a new [Atomic] for writing. Writes to it will be unbuffered. It will be
+// backed by a hidden (i.e. its name will start with a dot), temporary file (it
+// will have a .tmp extension). If the returned Atomic gets closed without
+// calling [Atomic.Finish] before, the temporary file will be deleted without
+// touching the target. Use like so:
+//
+//	f, err := file.AtomicWithTarget("foo").Open()
+//	if err != nil {
+//		return err
+//	}
+//	defer f.Close()
+//	_, err = f.Write([]byte("I am atomic!"))
+//	if err != nil {
+//		return err
+//	}
+//	return f.Finish()
+func (o *AtomicOpener) Open() (f *Atomic, err error) {
+	f = &Atomic{o.atomicOpts, nil}
+
+	// Determine the absolute path of the target. This is a safeguard to make
+	// the Atomic robust against intermediary working directory changes.
+	if f.target, err = filepath.Abs(f.target); err != nil {
+		return nil, err
+	}
+
+	// This will actually open the file in read/write mode,
+	// but we're not going to tell anyone about it.
+	f.fd, err = os.CreateTemp(filepath.Dir(f.target), fmt.Sprintf(".%s.*.tmp", filepath.Base(f.target)))
+	if err != nil {
+		return nil, err
+	}
+
+	var opened bool
+	defer func() {
+		if !opened {
+			err = errors.Join(err, f.fd.Close(), remove(f.fd))
```

Comment:
I'm not sure if these errors should be returned, any errors from trying to close or remove a broken fd here are not why the open failed.

Maybe just something like `_ = f.fd.Close(); _ = remove(f.fd)` before returning the error below.

Reply:
Maybe the variable name `open` is not chosen wisely. Perhaps `success` would be better. In fact, at this point in the code, the file has already been successfully opened, but we want to do something with it, i.e. change the permissions before writing the contents. This defer basically ensures that if subsequent operations fail before we return to the caller, we still clean everything up.

If the file opening itself had failed, it would indeed be useless to close/remove the file.

Reply:
Maybe that whole chmod business after CreateTemp is futile anyways. Need to check.

Reply:
Looks like it is. https://cs.opensource.google/go/go/+/refs/tags/go1.22.3:src/os/tempfile.go;l=33

```go
...
	for {
		name := prefix + nextRandom() + suffix
		f, err := OpenFile(name, O_RDWR|O_CREATE|O_EXCL, 0600)
...
```

Reply:
Killed with :fire: